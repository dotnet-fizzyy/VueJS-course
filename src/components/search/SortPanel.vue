<template>
    <div :class="rootStyle">
        <span :class="titleStyle">{{ availableItemsCount }} movies found</span>

        <Options
            label="Sort By"
            :options="options"
            :selected-option="selectedOption"
            @on-select-option="onSelectOption"
        />
    </div>
</template>

<script lang="ts">
import Options, { OptionGroup } from '@/components/common/Options.vue';
import Vue from 'vue';

const AvailableOptionGroups: OptionGroup[] = [
    { label: 'Release Date', name: 'rd' },
    { label: 'Rating', name: 'rt' },
];

export interface SortPanelProps {
    availableItemsCount: number;
}

export default Vue.extend({
    name: 'SortPanel',
    components: { Options },
    props: {
        availableItemsCount: {
            type: Number,
            required: true,
        },
    },
    data() {
        return {
            options: AvailableOptionGroups,
            selectedOption: AvailableOptionGroups[0].name,
        };
    },
    computed: {
        rootStyle(): string {
            return this.$style.root;
        },
        titleStyle(): string {
            return this.$style.title;
        },
    },
    methods: {
        onSelectOption(option: string): void {
            this.selectedOption = option;
        },
    },
});
</script>

<style lang="scss" module scoped>
@import '../../styles/fonts';

.root {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.title {
    @extend .font;

    color: $white;
    font-weight: bold;
    font-size: 16px;
}
</style>
