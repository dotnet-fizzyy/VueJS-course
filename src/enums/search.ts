export enum SearchByOptionNames {
    Title = 'title',
    Genre = 'genres',
}

export enum SortByOptionsNames {
    ReleaseDate = 'release_date',
    Rating = 'rating',
}

export enum SearchQueryParams {
    SearchTerm = 'term',
    SearchByOption = 'searchBy',
    SortByOption = 'sortBy',
}
