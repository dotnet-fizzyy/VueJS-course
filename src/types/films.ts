export interface Film {
    id: string;
    name: string;
    releaseYear: number;
    posterUrl?: string;
    posterAlt?: string;
    shortDescription: string;
}

export interface FilmShortDescription extends Film {}

export interface FilmFullDescription extends Film {
    rating: number;
    runtime: number;
    fullDescription: string;
}
