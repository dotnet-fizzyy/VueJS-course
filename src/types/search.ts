import { SearchQueryParams } from '@/enums/search';

export interface SearchParams {
    [SearchQueryParams.SearchTerm]?: string;
    [SearchQueryParams.SearchByOption]: string;
    [SearchQueryParams.SortByOption]: string;
}
