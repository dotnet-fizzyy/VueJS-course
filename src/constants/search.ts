import { ButtonGroupOption } from '@/types/button';
import { SearchByOptionNames, SortByOptionsNames } from '@/enums/search';

export const SortByOptions: ButtonGroupOption[] = [
    { label: 'Release Date', name: SortByOptionsNames.ReleaseDate },
    { label: 'Rating', name: SortByOptionsNames.Rating },
];

export const SearchByOptions: ButtonGroupOption[] = [
    { label: 'Title', name: SearchByOptionNames.Title },
    { label: 'Genre', name: SearchByOptionNames.Genre },
];
