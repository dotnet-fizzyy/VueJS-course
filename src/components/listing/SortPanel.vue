<template>
    <div :class="$style.root">
        <span :class="$style.title">{{ availableItemsCount }} movies found</span>

        <button-group
            label="Sort By"
            :options="options"
            :selected-option="selectedOption"
            @select-option="onSelectOption"
        />
    </div>
</template>

<script lang="ts">
import ButtonGroup from '@/components/common/ButtonGroup.vue';
import Vue from 'vue';
import { SortByOptions } from '@/constants/search';

export interface SortPanelProps {
    availableItemsCount: number;
}

export default Vue.extend({
    name: 'SortPanel',
    components: { ButtonGroup },
    props: {
        availableItemsCount: {
            type: Number,
            required: true,
        },
    },
    data() {
        return {
            options: SortByOptions,
            selectedOption: SortByOptions[0].name,
        };
    },
    methods: {
        onSelectOption(option: string): void {
            this.selectedOption = option;
        },
    },
});
</script>

<style lang="scss" module scoped>
@import '../../styles/fonts';

.root {
    width: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    box-sizing: border-box;
}

.title {
    @extend .font;

    color: $white;
    font-weight: bold;
    font-size: 16px;
}
</style>
