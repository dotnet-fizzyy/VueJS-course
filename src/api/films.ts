import Vue from 'vue';
import VueResource from 'vue-resource';
import { CollectionResponse, Film, FilmResponse } from '@/types/films';
import { mapFilmApiResponseToFilm } from '@/utils/films';

export default class FilmsApi {
    private static readonly Headers: Record<string, string> = {
        Accept: 'application/json',
        'Content-Type': 'application/json',
    };

    public static async getFilms(): Promise<CollectionResponse<Film>> {
        const response: VueResource.HttpResponse = await Vue.http.get(`${FilmsApi.getBaseUrl()}/movies`, {
            headers: FilmsApi.Headers,
        });
        const responseBody: CollectionResponse<FilmResponse> = await response.json();

        return {
            total: responseBody.total,
            limit: responseBody.limit,
            offset: responseBody.offset,
            data: responseBody.data.map(mapFilmApiResponseToFilm),
        } as CollectionResponse<Film>;
    }

    public static async getFilmById(id: string): Promise<Film> {
        const response: VueResource.HttpResponse = await Vue.http.get(`${FilmsApi.getBaseUrl()}/movies/${id}`, {
            headers: FilmsApi.Headers,
        });
        const responseBody: FilmResponse = await response.json();

        return mapFilmApiResponseToFilm(responseBody) as Film;
    }

    private static getBaseUrl(): string {
        const baseUrl = Vue.http?.options?.root;

        if (!baseUrl) {
            throw new Error('Base URL for communicating with API is not defined');
        }

        return baseUrl;
    }
}
