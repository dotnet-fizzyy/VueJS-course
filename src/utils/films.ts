import { Film, FilmPreview, FilmResponse } from '@/types/films';
import { castToString, getYearFromStringDate } from '@/utils/index';

export const mapFilmToFilmPreview = (film: Film): FilmPreview => ({
    id: film.id,
    name: film.name,
    releaseYear: film.releaseYear,
    posterUrl: film.posterUrl,
    posterAlt: film.posterAlt,
    shortDescription: film.shortDescription,
});

export const mapFilmApiResponseToFilm = (filmRes: FilmResponse): Film => ({
    id: castToString(filmRes.id),
    name: filmRes.title,
    releaseYear: getYearFromStringDate(filmRes.release_date),
    posterUrl: filmRes.poster_path,
    posterAlt: filmRes.title.toLowerCase(),
    shortDescription: filmRes.label ?? 'Unspecified',
    fullDescription: filmRes.overview ?? 'Unspecified',
    rating: filmRes.vote_average ?? 0,
    runtime: filmRes.runtime ?? 0,
    genres: filmRes.genres,
});
