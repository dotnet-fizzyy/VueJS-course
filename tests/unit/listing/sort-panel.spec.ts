import SortPanel from '@/components/listing/SortPanel.vue';
import Vue from 'vue';
import Vuex from 'vuex';
import { FilmActionTypes, changeSortByActionPayload } from '@/vuex/modules/films/actions';
import { FilmGetterProps } from '@/vuex/modules/films/getters';
import { SortByOptionsNames } from '@/enums/search';
import { Wrapper, shallowMount } from '@vue/test-utils';
import { createDefaultVueInstance } from '../../setup';
import { getFilmModuleType } from '@/vuex/store/utils';

const localVue = createDefaultVueInstance();

describe('SortPanel.vue Tests', () => {
    const availableItemsCount: number = 5;

    let actions;
    let getters;
    let store;

    beforeEach(() => {
        actions = {
            [getFilmModuleType(FilmActionTypes.ChangeSortBy)]: jest.fn(),
        };
        getters = {
            [getFilmModuleType(FilmGetterProps.SortBy)]: (): SortByOptionsNames => SortByOptionsNames.ReleaseDate,
            [getFilmModuleType(FilmGetterProps.AvailableCount)]: (): number => availableItemsCount,
        };

        store = new Vuex.Store({
            actions,
            getters,
        });

        store.dispatch = jest.fn();
    });

    const createComponent = (): Wrapper<Vue> =>
        shallowMount(SortPanel, {
            store,
            localVue,
        });

    const findAvailableItemsCountLabel = (wrapper: Wrapper<Vue>) => wrapper.find('[data-aqa-available-items-count]');

    const findButtonGroupComponent = (wrapper: Wrapper<Vue>) => wrapper.findComponent({ name: 'button-group' });

    it('Should render', () => {
        //Arrange
        const wrapper = createComponent();

        //Act & Assert
        expect(wrapper.element).toMatchSnapshot();
    });

    it('Should render exact "availableItemsCount" passed from getter', () => {
        //Arrange
        const wrapper = createComponent();

        //Act & Assert
        const availableItemsCountLabelWrapper = findAvailableItemsCountLabel(wrapper);

        expect(availableItemsCountLabelWrapper.text()).toEqual(`${availableItemsCount} movies found`);
    });

    it('Should dispatch action on select different sort option', () => {
        //Arrange
        const sortOptionEventName: string = 'select-option';

        const wrapper = createComponent();

        const expectedOption: string = 'different_sort_opt';

        //Act & Assert
        const sortOptionsWrapper = findButtonGroupComponent(wrapper);

        expect(sortOptionsWrapper.element).toBeTruthy();

        sortOptionsWrapper.vm.$emit(sortOptionEventName, expectedOption);

        expect(sortOptionsWrapper.emitted()[sortOptionEventName]).toBeTruthy();
        expect(store.dispatch).toHaveBeenCalledWith(changeSortByActionPayload(expectedOption as SortByOptionsNames));
    });
});
