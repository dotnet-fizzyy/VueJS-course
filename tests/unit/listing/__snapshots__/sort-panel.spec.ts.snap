// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortPanel.vue Tests Should render 1`] = `
<div
  class="root"
>
  <span
    class="title"
    id="available-items-count"
  >
    5 movies found
  </span>
   
  <button-group-stub
    label="Sort By"
    options="[object Object],[object Object]"
    selectedoption="release_date"
  />
</div>
`;
